File: scripts/eval.ts
/**
 * A/B evaluation of the identity layer.
 *
 *   npm run eval -- --persona <slug-or-id> [--n 6]
 *
 * For each out-of-knowledge question, generates an answer WITH the approved
 * profile and WITHOUT it (system_prompt only), then asks an LLM judge which one
 * better matches the persona's identity (voice + beliefs). Proves the profile
 * keeps the clone "in character" even when RAG has nothing to retrieve.
 */
import { loadEnvConfig } from '@next/env'

loadEnvConfig(process.cwd())

// Generic, persona-agnostic prompts unlikely to hit any knowledge base.
const QUESTIONS = [
  'Me dá um conselho rápido para começar bem a semana.',
  'O que você acha mais importante: talento ou disciplina? Por quê?',
  'Como você reage quando alguém discorda fortemente de você?',
  'Recomende um hábito simples que muda a vida de alguém.',
  'Qual a sua opinião sobre trabalhar nos fins de semana?',
  'Se tivesse que se descrever em uma frase, qual seria?',
]

async function main() {
  const { getPool, queryOne } = await import('../lib/db')
  const { buildSystemPrompt } = await import('../lib/prompt')
  const { chat, chatJSON } = await import('../lib/openai')
  type Persona = import('../lib/types').Persona

  const args = process.argv.slice(2)
  const ref = args[args.indexOf('--persona') + 1]
  const n = Number(args[args.indexOf('--n') + 1]) || 6
  if (!ref || ref.startsWith('--')) {
    console.error('Uso: npm run eval -- --persona <slug-ou-id> [--n 6]')
    process.exit(1)
  }

  const persona = await queryOne<Persona>(
    `select * from personas where slug = $1 or id::text = $1 limit 1`,
    [ref]
  )
  if (!persona) {
    console.error(`Persona não encontrada: ${ref}`)
    process.exit(1)
  }
  if (!persona.profile) {
    console.error(`A persona "${persona.name}" não tem perfil aprovado. Nada para comparar.`)
    process.exit(1)
  }

  const withProfileSystem = buildSystemPrompt(persona)
  const withoutProfileSystem = persona.system_prompt
  const questions = QUESTIONS.slice(0, n)

  let winsWith = 0
  let winsWithout = 0
  let ties = 0

  console.log(`Avaliando "${persona.name}" em ${questions.length} perguntas fora do RAG...\n`)

  for (const q of questions) {
    const [a, b] = await Promise.all([
      chat([{ role: 'system', content: withProfileSystem }, { role: 'user', content: q }]),
      chat([{ role: 'system', content: withoutProfileSystem }, { role: 'user', content: q }]),
    ])

    const verdict = await chatJSON<{ winner: 'A' | 'B' | 'tie'; reason: string }>(
      [
        {
          role: 'system',
          content:
            'Você é um juiz de fidelidade de persona. Dado o PERFIL de identidade e duas respostas (A e B) à mesma pergunta, decida qual reflete melhor a voz, as crenças e os limites do perfil. Responda JSON {"winner":"A"|"B"|"tie","reason":"..."}.',
        },
        {
          role: 'user',
          content: `PERFIL:\n${JSON.stringify(persona.profile)}\n\nPERGUNTA: ${q}\n\nRESPOSTA A:\n${a}\n\nRESPOSTA B:\n${b}`,
        },
      ],
      { temperature: 0 }
    )

    if (verdict.winner === 'A') winsWith++
    else if (verdict.winner === 'B') winsWithout++
    else ties++

    console.log(`• ${q}\n  → ${verdict.winner === 'A' ? 'COM perfil' : verdict.winner === 'B' ? 'SEM perfil' : 'empate'} — ${verdict.reason}\n`)
  }

  console.log('================ Resultado ================')
  console.log(`COM perfil:  ${winsWith}`)
  console.log(`SEM perfil:  ${winsWithout}`)
  console.log(`Empates:     ${ties}`)
  console.log('==========================================')

  await getPool().end()
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})

File: backend/app/cli/eval.py
"""A/B evaluation of the identity layer. Port of scripts/eval.ts.

  python -m app.cli.eval --persona <slug-or-id> [--n 6]

For each out-of-knowledge question, generates an answer WITH the approved profile
and WITHOUT it (system_prompt only), then asks an LLM judge which better matches
the persona's identity. Proves the profile keeps the clone "in character".
"""
from __future__ import annotations

import asyncio
import json

import typer

app = typer.Typer(add_completion=False)

# Generic, persona-agnostic prompts unlikely to hit any knowledge base.
QUESTIONS = [
    "Me dá um conselho rápido para começar bem a semana.",
    "O que você acha mais importante: talento ou disciplina? Por quê?",
    "Como você reage quando alguém discorda fortemente de você?",
    "Recomende um hábito simples que muda a vida de alguém.",
    "Qual a sua opinião sobre trabalhar nos fins de semana?",
    "Se tivesse que se descrever em uma frase, qual seria?",
]


async def _run(ref: str, n: int) -> None:
    from ..db import close_pool, query_one
    from ..llm import chat, chat_json
    from ..services.prompts import build_system_prompt

    persona = await query_one(
        "select * from personas where slug = $1 or id::text = $1 limit 1", [ref]
    )
    if not persona:
        typer.echo(f"Persona não encontrada: {ref}", err=True)
        raise typer.Exit(1)
    if not persona.get("profile"):
        typer.echo(f'A persona "{persona["name"]}" não tem perfil aprovado. Nada para comparar.', err=True)
        raise typer.Exit(1)

    with_profile_system = build_system_prompt(persona)
    without_profile_system = persona["system_prompt"]
    questions = QUESTIONS[:n]

    wins_with = wins_without = ties = 0
    typer.echo(f'Avaliando "{persona["name"]}" em {len(questions)} perguntas fora do RAG...\n')

    for q in questions:
        a, b = await asyncio.gather(
            chat([{"role": "system", "content": with_profile_system}, {"role": "user", "content": q}]),
            chat([{"role": "system", "content": without_profile_system}, {"role": "user", "content": q}]),
        )
        verdict = await chat_json(
            [
                {
                    "role": "system",
                    "content": 'Você é um juiz de fidelidade de persona. Dado o PERFIL de identidade e duas respostas (A e B) à mesma pergunta, decida qual reflete melhor a voz, as crenças e os limites do perfil. Responda JSON {"winner":"A"|"B"|"tie","reason":"..."}.',
                },
                {
                    "role": "user",
                    "content": f"PERFIL:\n{json.dumps(persona['profile'], ensure_ascii=False)}\n\nPERGUNTA: {q}\n\nRESPOSTA A:\n{a}\n\nRESPOSTA B:\n{b}",
                },
            ],
            temperature=0,
        )
        winner = verdict.get("winner")
        if winner == "A":
            wins_with += 1
        elif winner == "B":
            wins_without += 1
        else:
            ties += 1
        label = "COM perfil" if winner == "A" else "SEM perfil" if winner == "B" else "empate"
        typer.echo(f"• {q}\n  → {label} — {verdict.get('reason')}\n")

    typer.echo("================ Resultado ================")
    typer.echo(f"COM perfil:  {wins_with}")
    typer.echo(f"SEM perfil:  {wins_without}")
    typer.echo(f"Empates:     {ties}")
    typer.echo("==========================================")
    await close_pool()


@app.command()
def main(
    persona: str = typer.Option(..., "--persona", help="slug ou id da persona."),
    n: int = typer.Option(6, "--n", help="número de perguntas."),
) -> None:
    asyncio.run(_run(persona, n))


if __name__ == "__main__":
    app()
